src/app/study-filters/study-filters.component.ts

```diff
@@ -0,0 +1,101 @@
+import { Validate } from 'class-validator';
+import { DatasetNode } from 'app/dataset-node/dataset-node';
+import { DatasetsService } from 'app/datasets/datasets.service';
+import { findNodeById } from 'app/treelist-checkbox/treelist-checkbox.component';
+import { Subscription } from 'rxjs/internal/Subscription';
+
+@Component({
+  selector: 'gpf-study-filters-block',
+  templateUrl: './study-filters.component.html',
+  styleUrls: ['./study-filters.component.css']
+})
+export class StudyFiltersComponent extends StatefulComponent implements OnInit, OnDestroy {
+  @Input() public dataset: Dataset;
+
+  @Validate(SetNotEmpty, {message: 'Select at least one.'})
+  public selectedStudies: Set<string> = new Set<string>(['']); // Used by the child component(Treelist checkbox)
+  public allStudyNodes: DatasetNode[] = null;
+  public datasetSubscription: Subscription;
+
+  @ViewChild('nav') public ngbNav: NgbNav;
+
+  public constructor(protected store: Store, public datasets: DatasetsService) {
+    super(store, StudyFiltersState, 'studyFilters');
+  }
+
+  public showStudyFilters = false;
+
+  public ngOnInit(): void {
+    super.ngOnInit();
+    const datasets$ = this.datasets.getDatasetsObservable();
+    const datasetTrees = new Array<DatasetNode>();
+    this.datasetSubscription = datasets$.subscribe(datasets => {
+      datasets.filter(d => !d.parents.length).map(d => {
+        datasetTrees.push(new DatasetNode(d, datasets));
+      });
+      this.allStudyNodes = this.getAllStudyNodesById(datasetTrees, this.dataset.id);
+    });
+    this.store.selectOnce(state => state.studyFiltersState as StudyFiltersModel).subscribe(state => {
+      // restore state
+      if ((state?.studyFilters?.length ?? 0) !== 0) {
+        const timer = setTimeout(() => {
+          if (this.showStudyFilters === false) {
+            this.showStudyFilters = true;
+            this.ngbNav.select('studies');
+            this.updateSelectedStudies(new Set(state.studyFilters));
+            clearInterval(timer);
+          }
+        }, 200);
+      }
+    });
+  }
+
+  public ngOnDestroy(): void {
+    this.datasetSubscription.unsubscribe();
+  }
+
+  public onNavChange(): void {
+    if (this.ngbNav.activeId === 'studies') {
+      this.selectedStudies.add(''); // Otherwise the selected studies are going to be empty, preventing navigation
+    } else {
+      this.selectedStudies.clear();
+    }
+    this.store.dispatch(new SetStudyFilters([]));
+    this.store.dispatch(new StateReset(StudyFiltersState));
+  }
+
+  private getAllStudyNodesById(datasetTrees: DatasetNode[], datasetId: string): DatasetNode[] {
+    const result: DatasetNode[] = [];
+    for (const dataset of datasetTrees) {
+      if (dataset.dataset.id === datasetId) {
+        result.push(dataset);
+      } else if (dataset.children && dataset.children.length > 0) {
+        result.push(...this.getAllStudyNodesById(dataset.children, datasetId));
+      }
+    }
+    return result;
+  }
```

Comment:
The result of this method, `allStudyNodes` is only ever used for its first element. In fact, since datasets have unique IDs, the return value of this method should always be a single `DatasetNode` instance. Refactor this method so that it returns a single `DatasetNode` and rename it accordingly.

Reply:
Going even further, you already have a method that does this - `findNodeById`. Simply use it instead and completely remove this method. After that, rename the variable `allStudyNodes` to something more accurate, since it will no longer be an array.